### PR TITLE
Artifacts as Reward for Society Quests

### PR DESCRIPTION
**Pull request by MemeWizard**
#### Changelog:
- Artifact gamerules now belong to separate group.
- Added chance to receive generic artifact after completing a society quest. Added society artifact rewards gamerule to regulate this.
- Characters with higher title tiers will have increased chance to receive high tier artifacts, while characters with lower title tiers will have increased chance to receive low tier artifacts on gamestart. Added gamestart artifacts rarity gamerule to regulate this.
- Various factors, like culture, affect chances of getting specific types of artifacts. Added artifacts preferences gamerule to regulate this.

#### Developer changelog:
- Reworked effect that spawns artifact.
- Renamed RULE_INVADERS_STRENGTH text lines in localization file, since it used not only for that gamerule

#### How to test:
- Added `test.44`&`test.45` events. You can use options tooltips to test artifact rarity and preferences.
- You also may want to check [design document](https://docs.google.com/spreadsheets/d/1x0bDnGuc2NHGw2iRAzru9CxtG8S-3nIZZdnBTc0Us8A/edit#gid=1500269219&range=1:24)

### CODE REVIEW

File: common/scripted_triggers/wc_scripted_artifacts_triggers.txt [hunk 1]
```diff
@@ -28,4 +30,155 @@ crossbow_artifact_active_trigger = {
...
+	# }
+}
+
+prefers_twohanded_weapons_trigger = {
+	# OR = {
+		is_large_humanoid_trigger = yes
+	# }
+}
+
+prefers_swords_trigger = {
```
Comment (zumbak04) [inline common/scripted_triggers/wc_scripted_artifacts_triggers.txt:110]: There's `use_scimitars_trigger` and BASICALLY these cultures use swords.

Comment (MemeWizard) [reply common/scripted_triggers/wc_scripted_artifacts_triggers.txt:110]: Done

File: common/scripted_score_values/wc_scripted_generic_artifact_spawn_values.txt [hunk 2]
```diff
@@ -0,0 +1,774 @@
...
+		NOT = { has_game_rule = { name = artifact_spawn_rarity value = random } }
+		tier = count
+	}
+}
+generic_artifact_rarity_epic_score = {
+	modifier = { # In-game chances to spawn
+		factor = 2
+
+		check_variable = { which = local_new_generic_artifact_gamestart_variable value < 1 }
+		check_variable = { which = local_new_generic_artifact_gamestart_variable value < 1 }
```
Comment (zumbak04) [inline common/scripted_score_values/wc_scripted_generic_artifact_spawn_values.txt:53]: The same variable checked twice? What it's a point?

Comment (MemeWizard) [reply common/scripted_score_values/wc_scripted_generic_artifact_spawn_values.txt:53]: Probably forgot to remove it after changing variables
Will do it later

File: common/scripted_effects/wc_artifacts_generic_scripted_effects.txt [hunk 3]
```diff
@@ -735,248 +735,3119 @@ give_new_artifact_random_name_and_origin_effect = {
...
-						8 = { add_artifact = amulet_2_diplomacy }
+		3 = { # Jewelry
+			trigger = {
+
+				check_variable = { which = local_new_received_generic_artifact_type_jewelry_amount_variable value < 3 } # Because 3 is max, since there is only 3 jewelry types
+
+				#trigger_if = {
+					#limit = { check_variable = { which = local_new_generic_artifact_receiving_character_matter_variable value = 1 } }
+					trigger_if = {
+						limit = { check_variable = { which = local_new_generic_artifact_gamestart_variable value = 1 } }
```
Comment (zumbak04) [inline common/scripted_effects/wc_artifacts_generic_scripted_effects.txt:785]: This is supposed to be boolean, yes? Thus it has either 0 or 1?

Comment (MemeWizard) [reply common/scripted_effects/wc_artifacts_generic_scripted_effects.txt:785]: > This is supposed to be boolean, yes? Thus it has either 0 or 1?

Yeah.

File: common/scripted_score_values/wc_scripted_generic_artifact_spawn_values.txt (hunk 2 above)
Comment (MemeWizard) [reply common/scripted_score_values/wc_scripted_generic_artifact_spawn_values.txt:53]: Done

**Comment by zumbak04** [comment]
It's strange I get artefacts I can't use. It's especially sad if the character is the AI who pick a reward randomly.
<details>
<summary>Click to expand</summary>

image

</details>

File: localisation/00_memewc_united_localisation.csv [hunk 4]
```diff
@@ -9,6 +11,35 @@ RULE_ARTIFACT_SPAWN_X0_25_DESC;Amount of generic artifacts spawned on gamestart
...
 ;;;;;;;;;;;;;;x
+RULE_ARTIFACT_SPAWN_SOCIETY;�RWarCraft�!: Artifacts as Society Quests Rewards;;�RWarCraft�!: Artefakte als Belohnung f�r Aufgaben der Gesellschaft;;;;;;;;;;;x
+RULE_ARTIFACT_SPAWN_SOCIETY_DEFAULT_DESC;Default chances to get artifact after completing society quest.;;Standard-Chancen, ein Artefakt nach Abschluss einer Aufgabe der Gesellschaft zu erhalten.;;;;;;;;;;;x
+RULE_ARTIFACT_SPAWN_SOCIETY_X1_5_DESC;Chances to get artifact after completing society quest is modified by �G1.5�!;;Die Chancen, ein Artefakt nach Abschluss einer Aufgabe der Gesellschaft zu erhalten, werden um �G1,5�! erh�ht.;;;;;;;;;;;x
+RULE_ARTIFACT_SPAWN_SOCIETY_X2_DESC;Chances to get artifact after completing society quest is modified by �G2�!;;Die Chancen, ein Artefakt nach Abschluss einer Aufgabe der Gesellschaft zu erhalten, werden um �G2�! erh�ht.;;;;;;;;;;;x
+RULE_ARTIFACT_SPAWN_SOCIETY_X0_5_DESC;Chances to get artifact after completing society quest is modified by �R0.5�!;;Die Chancen, ein Artefakt nach Abschluss einer Aufgabe der Gesellschaft zu erhalten, werden um �R0,5�! reduziert.;;;;;;;;;;;x
```
Comment (zumbak04) [inline localisation/00_memewc_united_localisation.csv:18]: Dots.
```suggestion
RULE_ARTIFACT_SPAWN_SOCIETY_X1_5_DESC;Chances to get artifact after completing society quest is modified by �G1.5�!.;;Die Chancen, ein Artefakt nach Abschluss einer Aufgabe der Gesellschaft zu erhalten, werden um �G1,5�! erh�ht.;;;;;;;;;;;x
RULE_ARTIFACT_SPAWN_SOCIETY_X2_DESC;Chances to get artifact after completing society quest is modified by �G2�!.;;Die Chancen, ein Artefakt nach Abschluss einer Aufgabe der Gesellschaft zu erhalten, werden um �G2�! erh�ht.;;;;;;;;;;;x
RULE_ARTIFACT_SPAWN_SOCIETY_X0_5_DESC;Chances to get artifact after completing society quest is modified by �R0.5�!.;;Die Chancen, ein Artefakt nach Abschluss einer Aufgabe der Gesellschaft zu erhalten, werden um �R0,5�! reduziert.;;;;;;;;;;;x
```

Comment (MemeWizard) [reply localisation/00_memewc_united_localisation.csv:18]: done

**Comment by MemeWizard** [comment]
> It's strange I get artefacts I can't use. It's especially sad if the character is the AI who pick a reward randomly.

Sad, but I think the most obvious and easy way to fix it is to add more artifacts.
Otherwise, a severe code changes for event are required, and since these code changes become obsolete once more artifacts will be added, I find it impractical to write such code that will be deleted later anyway.

So, here is my solution:
I will block getting artifacts that character can't use, but decrease choice options from 4 to 3 until I add more artifacts

**Comment by zumbak04** [comment]
> I will block getting artifacts that character can't use, but decrease choice options from 4 to 3 until I add more artifacts

I'm fine.

**Review by Myrran — COMMENTED** [review]
Bugs:

- According to event test.44 all title tiers receive baron title chances of artifact spawn at game start regardless of “Rarity” game rule option chosen;
- According to event test.44 it seems that those races/cultures who have weapon preferences, have increased chances of receiving TH and OH swords, although they shouldn't (centaur, naga, elf). Tolvir, who doesn't have any preferences are not affected by this bug.

ck2_2

I haven't tested artifact rewards for society quests yet.

**Comment by MemeWizard** [comment]
> According to event test.44 all title tiers receive baron title chances of artifact spawn at game start regardless of “Rarity” game rule option chosen;

Seems like because I changed something, the code itself worked as intended, but event to test the code didn't work as intended, sorry bout that.

<details>
<summary>Should work normally now</summary>

image
image

</details>

> According to event test.44 it seems that those races/cultures who have weapon preferences, have increased chances of receiving TH and OH swords, although they shouldn't (centaur, naga, elf). Tolvir, who doesn't have any preferences are not affected by this bug. Tested using "Even stronger preferences" game rule option.

True, spreadsheets is wrong. Forgot to add cultures that uses scimitars to the list.

**Review by Myrran — COMMENTED** [review]
- Artifacts rarity works fine.

Bugs:

- event test.44 doesn’t show rifle preferences;
<details>
  <summary>Screenshot</summary>

  ck2_1
</details>


- Tirassian culture Old Ways and Gilnean Goldrinn doesn’t show weapon preferences at all;
<details>
  <summary>Screenshot</summary>

ck2_2
ck2_3
</details>


- Vrykul doesn’t have the same preferences for TH swords as for TH axes.
<details>
  <summary>Screenshot</summary>

ck2_4
</details>

**Comment by MemeWizard** [comment]
> event test.44 doesn’t show rifle preferences;

Characters with low techs can't get artifacts that require higher techs in order to be used.

<details>
<summary>Click to expand</summary>

image
image

</details>

> Tirassian culture Old Ways and Gilnean Goldrinn doesn’t show weapon preferences at all;

Druid is a magic class, and magic classes can't get weapons they cannot use - melee or ranged.

> Vrykul doesn’t have the same preferences for TH swords as for TH axes.

Fixed, should be ok now.

**Review by Myrran — COMMENTED** [review]
All right, artifact preferences seem to work fine.
Next step is to check that society quests give artifacts rewards. I'll check that tomorrow.

**Review by Myrran — APPROVED** [review]
Society quests do give artifacts rewards. Seems to work fine.

**Comment by MemeWizard** [comment]
@zumbak04 @arithon @Myrran thanks for reviews <3

**Pull request merged**